### PR TITLE
KAFKA-3519: Refactor Transformer's transform / punctuate to return nullable values

### PR DESCRIPTION
**Pull request by guozhangwang**

### CODE REVIEW

**Comment by guozhangwang** [comment]
Ping @apovzner @norwood @enothereska for reviews.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValuesTest.java [hunk 1]
```diff
@@ -58,7 +58,8 @@ public Integer transform(Integer value) {
                         }
 
                         @Override
-                        public void punctuate(long timestamp) {
+                        public Integer punctuate(long timestamp) {
+                            return null;
```
Comment (apovzner) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValuesTest.java]: Maybe add a unit test where we return non-null object from punctuate?

File: streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java [hunk 2]
```diff
@@ -44,17 +46,18 @@
...
     /**
-     * Perform any periodic operations, if this processor {@link ProcessorContext#schedule(long) schedule itself} with the context
+     * Perform any periodic operations and possibly generate a new key and value, if this processor {@link ProcessorContext#schedule(long) schedule itself} with the context
      * during {@link #init(ProcessorContext) initialization}.
      *
      * @param timestamp the stream time when this method is being called
+     * @return new value; if null no key-value pair will be forwarded to down stream
      */
-    void punctuate(long timestamp);
+    KeyValue<RK, RV> punctuate(long timestamp);
```
Comment (norwood) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java]: why don’t we return `Iterable<RK, RV>`? aren’t we limiting the punctuate to only return values for one key by doing this?

e.g., if i wanted `transform()` to do aggregation and only send values downstream on punctuate i'd have to do some pretty ugly flatmapping after this transform

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValuesTest.java (hunk 1 above)
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamTransformValuesTest.java]: Ack.

File: streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java (hunk 2 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java]: Hmm, that is a good point. Although I feel returning `Iterable<RK, RV>` be overkill for the special case here. Now I'm thinking about that, maybe we should just keep return type as `R` which can then be defined as `Iterable<RK, RV>`, and hence in your case can just be followed by a dummy flatMap?

Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java]: @norwood To be more concrete, I will revert the change to still use `R` as the return value, and you can then set `R`  As `KeyValue<Object, Interable<K, V>>` where the return key is always null, and follow the `transform` with a `flatMap((k, v) -> v)`. Would that work?

Comment (norwood) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java]: `transform` and `punctuate` will have the same return type?

Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Transformer.java]: Yes, I have just updated this PR. Please review and let me know your thoughts.

**Comment by norwood** [comment]
👍

**Comment by apovzner** [comment]
looks good to me

**Comment by guozhangwang** [comment]
Thanks for the reviews. Merged to trunk.

**Pull request closed without merging**